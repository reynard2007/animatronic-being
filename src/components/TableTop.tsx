import React, { useCallback, useState, useMemo } from 'react';
import classnames from 'classnames';
import { isNull } from 'util';

import Direction from '../models/Direction';
import Nullable from '../models/Nullable';

import Board from './Board';
import Robot from './robot/Robot';
import CommandInput from './CommandInput';
import styles from './TableTop.module.scss';
import { placementRegex } from './constants';
import TableTopContext, { TableTopContextValue } from './TableTopContext';

const directionRotationMap: Record<Direction, number> = {
  NORTH: 180,
  EAST: 270,
  WEST: 90,
  SOUTH: 0,
};

const TableTop = () => {
  const [position, setPosition] = useState<TableTopContextValue['position']>(
    null
  );
  const [reportedValue, setReportedValue] = useState('');
  const [reportVisible, setReportVisible] = useState(false);
  const [rotation, setRotation] = useState(0);
  // Using numeric values here to allow consecutive shakes
  const [restrictedCount, setRestrictedCount] = useState(0);

  const placeRobot = useCallback((command: string) => {
    const directionMatch = command.match(/(NORTH)|(EAST)|(WEST)|(SOUTH)$/);
    const positionMatch = command.match(/[0-4],[0-4]/);

    let coordinate: TableTopContextValue['position'] = null;
    if (directionMatch && positionMatch) {
      const [x, y] = positionMatch[0].split(',');
      coordinate = { x: parseInt(x), y: parseInt(y) };

      setRotation(directionRotationMap[directionMatch[0] as Direction]);
      setPosition(coordinate);
    }
  }, []);

  const direction = useMemo<Nullable<Direction>>(() => {
    const radians = (rotation * Math.PI) / 180;
    const calculatedX = Math.round(Math.cos(radians));
    const calculatedY = Math.round(Math.sin(radians));

    switch (true) {
      case calculatedX === 1 && calculatedY === 0:
        return 'SOUTH';
      case calculatedX === 0 && calculatedY === -1:
        return 'EAST';
      case calculatedX === -1 && calculatedY === 0:
        return 'NORTH';
      case calculatedX === 0 && calculatedY === 1:
        return 'WEST';
      default:
        return null;
    }
  }, [rotation]);

  const moveRobot = useCallback(() => {
    if (!position || isNull(rotation) || isNull(direction)) {
      return;
    }

    const { x, y } = position;

    switch (true) {
      case direction === 'NORTH' && y < 4:
        setPosition({ x, y: y + 1 });
        break;
      case direction === 'EAST' && x < 4:
        setPosition({ x: x + 1, y });
        break;
      case direction === 'WEST' && x > 0:
        setPosition({ x: x - 1, y });
        break;
      case direction === 'SOUTH' && y > 0:
        setPosition({ x, y: y - 1 });
        break;
      default:
        setRestrictedCount(val => val + 1);
    }
  }, [direction, position, rotation]);

  const runCommand = useCallback(
    (command: string) => {
      let reported = false;

      switch (true) {
        case placementRegex.test(command):
          placeRobot(command);
          break;
        case /^MOVE$/.test(command):
          moveRobot();
          break;
        case /^LEFT$/.test(command):
          setRotation(val => val - 90);
          break;
        case /^RIGHT$/.test(command):
          setRotation(val => val + 90);
          break;
        case /^REPORT$/.test(command):
          if (position && !isNull(rotation)) {
            setReportedValue(`${position.x},${position.y},${direction}`);
            setReportVisible(true);
            reported = true;
          }
          break;
        default:
      }

      if (!reported) {
        setReportVisible(false);
      }
    },
    [direction, moveRobot, placeRobot, position, rotation]
  );

  return (
    <TableTopContext.Provider
      value={{ rotation, position, runCommand, restrictedCount }}
    >
      <div className={styles['layout-wrapper']}>
        <div className={styles['board-wrapper']}>
          <div className={styles.centerer}>
            <Board />
            <Robot />

            {!isNull(rotation) && position && (
              <div
                className={classnames(styles.report, {
                  [styles.show]: reportVisible && reportedValue,
                })}
              >
                {reportedValue}
              </div>
            )}
          </div>
        </div>

        <div className={styles.commands}>
          <CommandInput />
        </div>
      </div>
    </TableTopContext.Provider>
  );
};

export default TableTop;
