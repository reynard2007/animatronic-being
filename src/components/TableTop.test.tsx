import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import TableTop from './TableTop';
import userEvent from '@testing-library/user-event';
import { act } from 'react-dom/test-utils';

describe('TableTop', () => {
  afterEach(() => {
    jest.clearAllTimers();
  });

  test('should render properly', () => {
    const {
      container,
      getByPlaceholderText,
      getByText,
      queryByLabelText,
    } = render(<TableTop />);

    expect(getByPlaceholderText('Commands go here')).toBeInTheDocument();
    expect(getByText('Execute')).toBeInTheDocument();
    expect(queryByLabelText('Toy robot')).not.toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  test('should show robot on PLACE', () => {
    const { getByPlaceholderText, getByText, getByLabelText } = render(
      <TableTop />
    );

    act(() => {
      userEvent.paste(
        getByPlaceholderText('Commands go here'),
        'PLACE 3,2,SOUTH'
      );
      fireEvent.click(getByText('Execute'));
    });

    expect(getByLabelText('Toy robot')).toBeInTheDocument();
  });

  describe('should report correct value', () => {
    test.each([
      [
        `PLACE 3,4,SOUTH
          RIGHT
          RIGHT
          RIGHT
          RIGHT
          MOVE
          movE
          mOvE
          move
          move
          Move
          MOVE
          movE
          mOvE
          move
          move
          Move
          REPORT`,
        '3,0,SOUTH',
      ],
      [
        `MOVE
          MOVE
          pLAce 3,0,soUTH
          LefT
          move
          move
          move
          move
          move
          MoVe
          REPORT`,
        '4,0,EAST',
      ],
      [
        `MOVE
          MOVE
          pLAce 3,0,soUTH
          LefT
          left
          LeFt
          move
          move
          move
          move
          move
          move
          move
          REPORT`,
        '0,0,WEST',
      ],
      [
        `MOVE
          MOVE
          pLAce 3,0,soUTH
          LefT
          LEFt
          move
          move
          move
          move
          move
          REPORT`,
        '3,4,NORTH',
      ],
      [
        `MOVE
          MOVE
          pLAce 3,0,soUTH
          PLACE 5,10,NORTH
          REPORT`,
        '3,0,SOUTH',
      ],
    ])('%s should call %s', (input, report) => {
      const { getByPlaceholderText, getByText } = render(<TableTop />);

      jest.useFakeTimers();
      act(() => {
        userEvent.paste(getByPlaceholderText('Commands go here'), input);
        fireEvent.click(getByText('Execute'));
      });

      jest.runAllTimers();
      expect(getByText(report)).toBeInTheDocument();
    });
  });
});
