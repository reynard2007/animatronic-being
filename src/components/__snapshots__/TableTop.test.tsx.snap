// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableTop should render properly 1`] = `
<div>
  <div
    class="layout-wrapper"
  >
    <div
      class="board-wrapper"
    >
      <div
        class="centerer"
      >
        <div
          class="board"
          role="img"
        >
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
          <div
            class="square"
            data-testid="square"
          />
        </div>
      </div>
    </div>
    <div
      class="commands"
    >
      <div
        class="flex"
      >
        <textarea
          class="commands"
          id="bot-commands"
          name="bot-commands"
          placeholder="Commands go here"
          rows="10"
        />
        <button
          class="button"
        >
          Execute
        </button>
      </div>
    </div>
  </div>
</div>
`;
